python/ql/src/experimental/Security/CWE-340/TokenBuiltFromUUID.ql

```diff
@@ -0,0 +1,69 @@
+/**
+ * @name Predictable token
+ * @description Tokens used for sensitive tasks, such as, password recovery,
+ *  and email confirmation, should not use predictable values.
+ * @kind path-problem
+ * @precision medium
+ * @problem.severity error
+ * @security-severity 5
+ * @id py/predictable-token
+ * @tags security
+ *       external/cwe/cwe-340
+ */
+
+import python
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.ApiGraphs
+import semmle.python.dataflow.new.TaintTracking
+import DataFlow::PathGraph
+
+class PredictableResultSource extends DataFlow::Node {
+  PredictableResultSource() {
+    exists(API::Node uuidCallRet |
+      uuidCallRet =
+        API::moduleImport("uuid")
+            .getMember(["uuid1", "uuid2", "uuid3", "uuid5"])
+            .getACall()
+            .getReturn()
+    |
+      this = uuidCallRet.asSource()
+      or
+      this = uuidCallRet.getMember(["hex", "bytes", "bytes_le"]).asSource()
+    )
+  }
+}
+
+class TokenAssignmentValueSink extends DataFlow::Node {
+  TokenAssignmentValueSink() {
+    exists(Assign a, Expr target | this = DataFlow::exprNode(a.getValue()) |
```

Comment:
For more generality, it would be better to use `DefinitionNode` here, and handle things at the control-flow layer. (This would make the code cover not only simple assignments, but also things like `:=` and `as`, which serve the same function, but are different at the level of syntax.)

Also, it would be better to use `AttrWrite` rather than `Attribute`. The former is defined at the data-flow layer, and again covers more constructs that all write to attributes.

Reply:
Could you elaborate on this a bit more? I understand the concepts of what you said, and it makes totally sense. However, I am not familiar with the CodeQL python API enough to re-write this according to your suggestions.

Thank you!

Reply:
Does this reflect your suggestions?

```codeql
class TokenAssignmentValueSink extends DataFlow::Node {
  TokenAssignmentValueSink() {
    exists(DefinitionNode dn, DataFlow::AttrWrite aw | this = aw.getValue() |
      aw.getValue().asCfgNode() = dn.getValue() and
      aw.getAttributeName().toLowerCase().matches(["%token", "%code"])
    )
  }
}
```

Reply:
Uow! Thanks for the class! I was always bothered by the string repetition but never thought about using a string variable the way you did.  Moreover, I was missing the cast to NameNode idea to make your suggestion. I went with the one with the outer and inner exists cause I do agree readability is better.